File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/between/SleepSyncTest.kt
package org.wyrdsekai.app.engine.between

import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class SleepSyncTest {

    @Test
    fun buildRequestIncludesAllFields() {
        val between = InMemoryBetweenClient()
        val manager = SleepSyncManager(between, "node-1", "family-1")

        val request = manager.buildRequest(
            budDid = "bud-1",
            manifestVersion = 3,
            localItemHashes = listOf("hash-a", "hash-b"),
            localTombstones = listOf(
                Tombstone("hash-c", "superseded", "bud-1", 100),
            ),
            lastSyncTimestamp = 5000,
        )

        assertEquals("bud-1", request.budDid)
        assertEquals("node-1", request.nodeId)
        assertEquals(3, request.manifestVersion)
        assertEquals(2, request.localItemHashes.size)
        assertEquals(1, request.localTombstones.size)
        assertEquals(5000, request.lastSyncTimestamp)
        assertTrue(request.timestamp > 0)
    }

    @Test
    fun requestSyncPublishesToBetween() = runTest {
        val between = InMemoryBetweenClient()
        between.connect("ws://test")
        val manager = SleepSyncManager(between, "node-1", "family-1")

        val request = SleepSyncRequest(
            budDid = "bud-1",
            nodeId = "node-1",
            manifestVersion = 1,
            localItemHashes = listOf("h1"),
            lastSyncTimestamp = 0,
            timestamp = 1000,
        )
        manager.requestSync(request)

        assertEquals(1, between.published.size)
        val subject = between.published[0].first
        assertTrue(subject.contains("soul.sync.request"))
        assertTrue(subject.contains("family-1"))
        assertTrue(subject.contains("node-1"))
    }

    @Test
    fun receiveSyncResponseTriggersCallback() = runTest {
        val between = InMemoryBetweenClient()
        between.connect("ws://test")
        val manager = SleepSyncManager(between, "node-1", "family-1")

        var received: SleepSyncResponse? = null
        manager.onSyncResponse { received = it }
        manager.startListening()

        // Simulate server response
        val response = SleepSyncResponse(
            budDid = "bud-1",
            newItems = listOf(
                SoulItemRef("hash-x", "memory", 0.9f, "bud-2", 500),
            ),
            newTombstones = listOf(
                Tombstone("hash-old", "expired", "system", 600),
            ),
            manifestUpdated = true,
            itemsMerged = 3,
            tombstonesApplied = 1,
            timestamp = 2000,
        )
        val data = kotlinx.serialization.json.Json.encodeToString(
            SleepSyncResponse.serializer(), response
        ).encodeToByteArray()
        between.publish("between.household.family-1.server.node-1.soul.sync.response", data)

        assertNotNull(received)
        assertEquals(1, received!!.newItems.size)
        assertEquals("hash-x", received!!.newItems[0].hash)
        assertEquals(true, received!!.manifestUpdated)
        assertEquals(3, received!!.itemsMerged)

        manager.stopListening()
    }

    @Test
    fun syncRequestSubjectMatchesPattern() = runTest {
        val between = InMemoryBetweenClient()
        between.connect("ws://test")
        val manager = SleepSyncManager(between, "node-1", "family-abc")

        val request = SleepSyncRequest(
            budDid = "bud-1", nodeId = "node-1",
            manifestVersion = 1, localItemHashes = emptyList(),
            lastSyncTimestamp = 0, timestamp = 100,
        )
        manager.requestSync(request)

        assertEquals(
            "between.household.family-abc.node-1.server.soul.sync.request",
            between.published[0].first,
        )
    }

    @Test
    fun tombstoneRoundTrip() {
        val json = kotlinx.serialization.json.Json
        val tombstone = Tombstone("hash-1", "expired", "bud-1", 1000)
        val encoded = json.encodeToString(Tombstone.serializer(), tombstone)
        val decoded = json.decodeFromString(Tombstone.serializer(), encoded)
        assertEquals(tombstone, decoded)
    }

    @Test
    fun soulItemRefRoundTrip() {
        val json = kotlinx.serialization.json.Json
        val ref = SoulItemRef("abc123", "personality", 0.95f, "bud-1", 500)
        val encoded = json.encodeToString(SoulItemRef.serializer(), ref)
        val decoded = json.decodeFromString(SoulItemRef.serializer(), encoded)
        assertEquals(ref, decoded)
    }
}

File: clients/rn/__tests__/engine/sleep-sync.test.ts
import { InMemoryBetweenClient } from '../../src/engine/between/BetweenClient';
import { SleepSyncManager, type SleepSyncRequest, type SleepSyncResponse } from '../../src/engine/between/SleepSync';

describe('SleepSyncManager', () => {
  it('buildRequest includes all fields', () => {
    const between = new InMemoryBetweenClient();
    const manager = new SleepSyncManager(between, 'node-1', 'family-1');

    const request = manager.buildRequest({
      budDid: 'bud-1',
      manifestVersion: 3,
      localItemHashes: ['hash-a', 'hash-b'],
      localTombstones: [
        { itemHash: 'hash-c', reason: 'superseded', createdBy: 'bud-1', timestamp: 100 },
      ],
      lastSyncTimestamp: 5000,
    });

    expect(request.budDid).toBe('bud-1');
    expect(request.nodeId).toBe('node-1');
    expect(request.manifestVersion).toBe(3);
    expect(request.localItemHashes.length).toBe(2);
    expect(request.localTombstones.length).toBe(1);
    expect(request.lastSyncTimestamp).toBe(5000);
    expect(request.timestamp).toBeGreaterThan(0);
  });

  it('requestSync publishes to Between', async () => {
    const between = new InMemoryBetweenClient();
    await between.connect('ws://test');
    const manager = new SleepSyncManager(between, 'node-1', 'family-1');

    const request: SleepSyncRequest = {
      budDid: 'bud-1',
      nodeId: 'node-1',
      manifestVersion: 1,
      localItemHashes: ['h1'],
      localTombstones: [],
      lastSyncTimestamp: 0,
      timestamp: 1000,
    };
    manager.requestSync(request);

    expect(between.published.length).toBe(1);
    const { subject } = between.published[0];
    expect(subject).toContain('soul.sync.request');
    expect(subject).toContain('family-1');
    expect(subject).toContain('node-1');
  });

  it('receive sync response triggers callback', async () => {
    const between = new InMemoryBetweenClient();
    await between.connect('ws://test');
    const manager = new SleepSyncManager(between, 'node-1', 'family-1');

    let received: SleepSyncResponse | null = null;
    manager.onSyncResponse(r => { received = r; });
    manager.startListening();

    const response: SleepSyncResponse = {
      budDid: 'bud-1',
      newItems: [
        { hash: 'hash-x', category: 'memory', significance: 0.9, createdBy: 'bud-2', timestamp: 500 },
      ],
      newTombstones: [
        { itemHash: 'hash-old', reason: 'expired', createdBy: 'system', timestamp: 600 },
      ],
      manifestUpdated: true,
      itemsMerged: 3,
      tombstonesApplied: 1,
      timestamp: 2000,
    };
    const data = new TextEncoder().encode(JSON.stringify(response));
    between.publish('between.household.family-1.server.node-1.soul.sync.response', data);

    expect(received).not.toBeNull();
    expect(received!.newItems.length).toBe(1);
    expect(received!.newItems[0].hash).toBe('hash-x');
    expect(received!.manifestUpdated).toBe(true);
    expect(received!.itemsMerged).toBe(3);

    manager.stopListening();
  });

  it('sync request subject matches pattern', async () => {
    const between = new InMemoryBetweenClient();
    await between.connect('ws://test');
    const manager = new SleepSyncManager(between, 'node-1', 'family-abc');

    const request: SleepSyncRequest = {
      budDid: 'bud-1', nodeId: 'node-1',
      manifestVersion: 1, localItemHashes: [],
      localTombstones: [], lastSyncTimestamp: 0, timestamp: 100,
    };
    manager.requestSync(request);

    expect(between.published[0].subject).toBe(
      'between.household.family-abc.node-1.server.soul.sync.request',
    );
  });
});
